lib/compiler/src/beam_doc.erl

```diff
@@ -0,0 +1,276 @@
+    catch E:R:ST ->
+            erlang:raise(E, R, ST)
+    end.
+
+extract_meta(AST, MetaData) ->
+    foldl(fun ({attribute, _ModuleDocAnno, moduledoc, MetaDoc}, Meta) when is_map(MetaDoc) ->
+                  maps:merge(Meta, MetaDoc);
+              (_, Meta) -> Meta
+          end, MetaData, AST).
+
+-spec extract_moduledoc(AST :: [tuple()]) -> ModuleDoc :: {erl_anno:anno(), binary() | none | hidden}.
+extract_moduledoc([]) ->
+    {?DEFAULT_MODULE_DOC_LOC, none};
+extract_moduledoc([{attribute, ModuleDocAnno, moduledoc, false}| _AST]) ->
+    extract_moduledoc([{attribute, ModuleDocAnno, moduledoc, hidden}| _AST]);
+extract_moduledoc([{attribute, ModuleDocAnno, moduledoc, hidden}| _AST]) ->
+    {ModuleDocAnno, hidden};
+extract_moduledoc([{attribute, ModuleDocAnno, moduledoc, ModuleDoc}| _AST]) when is_list(ModuleDoc) ->
+    {ModuleDocAnno, unicode:characters_to_binary(string:trim(ModuleDoc))};
+extract_moduledoc([_| AST]) ->
+    extract_moduledoc(AST).
+
+extract_docformat([]) ->
+    ?DEFAULT_FORMAT;
+extract_docformat([{attribute, _ModuleDocAnno, moduledoc, MetaFormat} | Ls]) when is_map(MetaFormat) ->
+    case maps:get(format, MetaFormat, not_found) of
+        not_found -> extract_docformat(Ls);
+        Format when is_list(Format) -> unicode:characters_to_binary(Format);
+        Format when is_binary(Format) -> Format
+    end;
+extract_docformat([_ | Ls]) ->
+    extract_docformat(Ls).
+
+-spec create_module_doc(ModuleDoc :: binary()) -> map().
+create_module_doc(ModuleDoc) when is_atom(ModuleDoc) ->
+    ModuleDoc;
+create_module_doc(ModuleDoc) when not is_atom(ModuleDoc) ->
+    create_module_doc(<<"en">>, ModuleDoc).
+
+-spec create_module_doc(Lang :: binary(), ModuleDoc :: binary()) -> map().
+create_module_doc(Lang, ModuleDoc) ->
+    #{Lang => ModuleDoc}.
+
+-spec new_state(Dirname :: unicode:chardata()) -> internal_docs().
+new_state(Dirname) ->
+    reset_state(#docs{cwd = Dirname}).
+
+-spec reset_state(State :: internal_docs()) -> internal_docs().
+reset_state(State) ->
+    State#docs{doc = none, meta = maps:new()}.
+
+-spec update_meta(State :: internal_docs(), Meta :: map()) -> internal_docs().
+update_meta(#docs{meta = Meta0}=State, Meta1) ->
+    State#docs{meta = maps:merge(Meta0, Meta1)}.
+
+-spec update_doc(State :: internal_docs(), Doc :: unicode:chardata() | atom()) -> internal_docs().
+update_doc(State, Doc) ->
+State#docs{doc = string:trim(Doc)}.
+
+-spec update_module(State :: internal_docs(), ModuleName :: unicode:chardata()) -> internal_docs().
+update_module(#docs{}=State, ModuleName) ->
+    State#docs{module_name = ModuleName}.
+
+-spec update_export_funs(State :: internal_docs(), proplists:proplist()) -> internal_docs().
+update_export_funs(State, ExportedFuns) ->
+    State#docs{exported_functions = sets:from_list(ExportedFuns)}.
+
+-spec update_export_types(State :: internal_docs(), proplists:proplist()) -> internal_docs().
+update_export_types(State, ExportedTypes) ->
```

Comment:
Yes, that makes sense.

Reply:
Given this reasoning of adding internal types to the beam chunk, the internal types will be added with documentation, if they have a doc attribute attached to them

Do we add internal types with documentation to the beam chunk but ignore private functions?
I ask because this is not consistent. We could add a new meta field that declares whether something is exported or not, as suggested, and let the IDEs decide to what show, in a consistent manner

Reply:
It will have to be inconsistent as types and functions are different. We do not want to export all types that are documented, as then they will become part of the API and their names need to be backward compatible. However, this means that there will be types that are public, but not exported, which is why types need to be treated differently than functions...

Maybe we will need to revisit this later, but I think for now this is the best approach.